### ISSUE TITLE
Please further clarify the difference in Modes

### ISSUE THREAD

**Issue by rplankenhorn** [seek-oss/github-merged-pr-buildkite-plugin#25]
The following is very confusing:

# Modes

The plugin has two modes:
  - `mode: checkout` to merge the PR after checking out source; and
  - `mode: trigger` to async trigger another build (of the current pipeline)

I would like to use this plugin to always merge the PR into the base branch before running our checks on Buildkite and I want new builds to trigger when the base branch changes.

1. Can I only use one mode?
2. If I use the ```trigger``` mode, will it also do the checkout for those steps?

**Comment by zsims** [comment]
Hi, thanks for raising an issue.

> I would like to use this plugin to always merge the PR into the base branch before running our checks 

In this case you'd use `mode: checkout` if you want to only have one build, and the merge would be done before your checks. `mode: trigger` can be used if you want to kick off a new build for "what the PR would look like if it was merged" -- giving you two builds, one on the unmerged PR and another on the merged PR.

> 1. Can I only use one mode?

Yes. Not sure how using both would help you in your case?

> 2. If I use the trigger mode, will it also do the checkout for those steps?

Yes, providing the plugin is used in each of the steps. As trigger works by:

Kicking off a new build with the SHA:

https://github.com/seek-oss/github-merged-pr-buildkite-plugin/blob/aaff3e4d3fcf22ad7a2012063d0005600f8b8f56/hooks/post-checkout#L107

https://github.com/seek-oss/github-merged-pr-buildkite-plugin/blob/aaff3e4d3fcf22ad7a2012063d0005600f8b8f56/hooks/post-checkout#L10-L20

Then when that step runs it will "force merge" the changes:

https://github.com/seek-oss/github-merged-pr-buildkite-plugin/blob/aaff3e4d3fcf22ad7a2012063d0005600f8b8f56/hooks/post-checkout#L89-L95

TL;DR we'd normally suggest using `merge` which gets you back on parity with the now defunct github `/merge` PR ref. And only use `trigger` if your use-case is sufficiently complicated.

You can currently only "trigger" PRs to re-build if you're using `mode: trigger`:
https://github.com/seek-oss/github-merged-pr-buildkite-plugin/blob/aaff3e4d3fcf22ad7a2012063d0005600f8b8f56/hooks/post-checkout#L114l-L119

Although we could probably change this PR to allow you to "re-trigger" PR builds when the base branch changes. Keen for your thoughts on this @seekdave ?

I think a lot of confusion has come from us somewhat overloading the term "trigger". To mean "triggering another build" but also "re-trigger the build of PRs"

**Comment by seekdave** [comment]
My team currently use the plugin in the way that is described by @rplankenhorn.

All of our main steps use the plugin in `checkout` mode, e.g.:
```
  - label: ":building_construction: Build"
    commands:
      - "make distclean dist"
    plugins:
      - seek-oss/github-merged-pr#v1.1.2:
          mode: checkout
```

Then we also include this PR triggering step to run on all `master` builds (master typically only builds after a merge):
```
  - label: ":github: Trigger PR branches"
    command: "true"
    branches: "master"
    plugins:
      - seek-oss/github-merged-pr#v1.1.2:
          mode: trigger
          update_prs: true
```

All PRs therefore build and test the merged result, and all builds on `master` re-trigger all open PRs. It was important for us to ensure we have a buildkite agent always ready and add the "Trigger PRs" at the start of the pipeline - this reduces the time window between a merge and triggering other PRs (to block them from merging until after their new build).

Note that the git-fu used to identify PR branches that need to be triggered requires to to ensure you clean up branches for closed (unmerged) PRs, or they will re-trigger and waste build resources.

Please let me know if that's unclear.

**Comment by rplankenhorn** [comment]
@seekdave @zsims Thanks!  This helps a lot!
